NODE-666: Deploy processing results RDBMS

igor-ramazanov opened the pull request:
### Overview
Stores deploy processing results into SQLite. Also, split `DeployBuffer` to separate reader/writer. And added new `status_message` column.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/NODE-666

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [x] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](http://drone.casperlabs.io/) system. This is necessary to run tests on this PR.

### Notes
* Please, pay attention that I've decided to store deploy processing results for all blocks, even invalid and not sure if it's a good idea.
* I'm still keeping `DeployStorageReader/Writer` under the `io.casperlabs.casper.deploybuffer` package, but will change it in another PR in the future.
* I won't merge it into `dev` branch, created a separate `NODE-659-rdbms-epic` branch for that.
* Suggest reviewing it commit-by-commit
* Don't be scared by a too large diff - they are mostly renamings

goral09 commented:
@igor-ramazanov just FYI, I assigned myself for the review since Akosh is on holidays :beach_umbrella: :tropical_drink: :pineapple:

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/Casper.scala:
```diff
@@ -91,7 +91,7 @@ sealed abstract class MultiParentCasperInstances {
 
   /** Create a MultiParentCasper instance from the new RPC style gossiping. */
-  def fromGossipServices[F[_]: Concurrent: Log: Time: Metrics: FinalityDetector: BlockStorage: DagStorage: ExecutionEngineService: LastFinalizedBlockHashContainer: DeployBuffer: Validation](
+  def fromGossipServices[F[_]: Concurrent: Log: Time: Metrics: FinalityDetector: BlockStorage: DagStorage: ExecutionEngineService: LastFinalizedBlockHashContainer: DeployStorageReader: DeployStorageWriter: Validation](
```
Maybe a simple `DeployStorage` trait could extend both, for convenience when no specialization is required? It could be used similarly, just `DeployStorage[F].readPending` and `DeployStorage[F].addAsPending` if your program needs to do both things anyway.

Similar to how for example Play JSON has a Reads, Writes and a Format which is both, and then `Json.write` is happy with a `Reads` or `Format` instance.

igor-ramazanov replied:
Done.

aakoshh replied:
Awesome, feels lighter, thanks.

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/deploybuffer/DeployStorageReader.scala:
```diff
@@ -0,0 +1,153 @@
+  // waiting to be finalized or orphaned
+  private val ProcessedStatusCode = 1
+  // Deploys that have been discarded for some reason and should be deleted after a while
+  // private val DiscardedStatusCode = 2
```
I don't think these constants should be duplicated. I suggest you either move them to a companion object, a base class or, what I think is the better, have it like this: 
```
@typeclass trait DeployStorageReader ...
@typeclass trait DeployStorageWriter
@typeclass trait DeployStorage extends DeployStorageReader with DeployStorageWriter
class SQLiteDeployStorage extends DeployStorage
```

igor-ramazanov replied:
Done

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/deploybuffer/DeployStorageReader.scala:
```diff
@@ -0,0 +1,153 @@
+  // private val DiscardedStatusCode = 2
+
+  private implicit val metaByteString: Meta[ByteString] =
+    Meta[Array[Byte]].imap(ByteString.copyFrom)(_.toByteArray)
```
This should be shared as well.

igor-ramazanov replied:
Done

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/deploybuffer/DeployStorageReader.scala:
```diff
@@ -0,0 +1,153 @@
+    val readProcessingResults =
+      sql"""|SELECT block_hash, cost, execution_error_message
+            |FROM deploys_process_results 
+            |WHERE deploy_hash=$hash""".stripMargin
```
It would be nice to order these by `execute_time_millis desc` so we can get a deterministic order and know which one is the "latest" result.

igor-ramazanov replied:
Done

aakoshh commented on casper/src/main/scala/io/casperlabs/casper/deploybuffer/DeployStorageWriter.scala:
```diff
@@ -92,12 +81,61 @@ class DeployBufferImpl[F[_]: Metrics: Time: Bracket[?[_], Throwable]](
+          |""".stripMargin
+      ).updateMany(
+        block.getBody.deploys
+          .map(
```
There should be a `zipWithIndex` here an the position of the result in `Block` body saved, so that we can restore the block precisely from these tables.

igor-ramazanov replied:
Nice catch! Done

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,36 @@
+    execute_time_millis     INTEGER NOT NULL,
+    cost                    INTEGER NOT NULL,
+    execution_error_message TEXT,
+    PRIMARY KEY (deploy_hash, block_hash),
```
I think the order of the primary keys should be `(block_hash, deploy_hash)` because most of the time we'd be looking for what's included in a block and insert them in a batch. This table I think is another good candidate for `without rowid`.

igor-ramazanov replied:
Done

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,36 @@
+);
+
+CREATE INDEX idx_deploys_process_results_account_create_time_deploy_hash ON deploys_process_results (account, create_time_millis, deploy_hash);
+CREATE INDEX idx_deploys_process_results_account_execute_time_deploy_hash ON deploys_process_results (account, create_time_millis, deploy_hash);
```
We already have an index on `deploys.account`, isn't that enough?

igor-ramazanov replied:
Maybe, I wanted to make it "covering index" for the query like `SELECT deploy_hash FROM deploys_process_results WHERE account=? and create_time_millis > ?`.

igor-ramazanov replied:
I've checked the query with `join` and it works well: 
```
sqlite> EXPLAIN QUERY PLAN SELECT deploy_hash FROM deploys INNER JOIN deploys_process_results dpr on deploys.hash = dpr.deploy_hash WHERE dpr.account='123' and execute_time_millis>0;
QUERY PLAN
|--SEARCH TABLE deploys_process_results AS dpr USING INDEX idx_deploys_process_results_account_execute_time_deploy_hash (account=?)
`--SEARCH TABLE deploys USING COVERING INDEX sqlite_autoindex_deploys_1 (hash=?)
```

igor-ramazanov replied:
Done (I'm still using indexes but only on timestamp columns)

aakoshh replied:
Right, I think these indexes can be removed then because as it indicates there's already a covering index in the `deploys` table, so we can grab the rows we are filtering for from there, and just join the processing results results, possibly joining further for the fault tolerance of the block, then aggregate to see if the whole deploy is ready or not. That's why you need the index on `deploy_process_results.deploy_hash`. 

That query you are explaining there: there is no reason for you to perform the join because every column you are filtering and projecting is present in the `deploy_process_results` table already. 

I suggest we add the indexes when we add the query to the program, it will be easier to see what's the access pattern. I'd either:
a) remove the indexes altogether and rely on the indexes that exist on `deploys` or
b) keep the `account` in the index as well as whatever else is needed, but only if we actually start utilising this

igor-ramazanov replied:
Done with the `a)` bullet

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,36 @@
+    cost                    INTEGER NOT NULL,
+    execution_error_message TEXT,
+    PRIMARY KEY (deploy_hash, block_hash),
+    FOREIGN KEY (deploy_hash) REFERENCES deploys (hash)
```
There should be an explicit index created for the foreign key too.

igor-ramazanov replied:
Done

igor-ramazanov commented:
@aakoshh Thanks for the thorough review! I've addressed the comments.

>The other thing is that I think we can try to just keep the Block Header in a table and restore the block from the processed deploys, but for that we need to be able to preserve the exact order of them in the body. It should help with orphans at least.

Yeah, that's a good idea. Though, I don't really know what's the best way to store block headers, because if I store them as a new column then it will be duplicated for all deploy process results which is highly inefficient from the storage space perspective.

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,39 @@
+
+CREATE INDEX idx_deploys_process_results_account_create_time ON deploys_process_results (create_time_millis);
+CREATE INDEX idx_deploys_process_results_account_execute_time ON deploys_process_results (execute_time_millis);
+CREATE INDEX idx_deploys_process_results_account_deploy_hash ON deploys_process_results (deploy_hash);
```
All these indexes have `account` in there name but now none of them have it in their field list.

igor-ramazanov replied:
Done, I've removed these indexes

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,39 @@
+) WITHOUT ROWID;
+
+CREATE INDEX idx_deploys_process_results_account_create_time ON deploys_process_results (create_time_millis);
+CREATE INDEX idx_deploys_process_results_account_execute_time ON deploys_process_results (execute_time_millis);
```
The only select I see in `SQLiteDeployStorage` is a `select * where deploy_hash = ? order by execute_time_millis desc`. That _could_ benefit from an index on `(deploy_hash, execute_time_millis desc)` but I think the hash is enough, usually there will be only one block, no need to store all the extra data in the index. But these indexes which only have the time in them, where are they used?

igor-ramazanov replied:
Currently, they're not used, except `getPendingOrProcessed(hash: ByteString)`

igor-ramazanov replied:
Ok, we can safely remove indexes and add them later when we have a cleaner use cases

igor-ramazanov replied:
Done

aakoshh replied:
I can't see it being used in `getPendingOrProcessed`. That one filters for a given deploy hash; the optimiser will try to pick the index it can use most effectively, and I don't think it can combine multiple indexes, one for filtering and another one for sorting. There are databases that can do such thing but I couldn't find any indication that SQLite is one of them.

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,39 @@
+    cost                    INTEGER NOT NULL,
+    execution_error_message TEXT,
+    PRIMARY KEY (block_hash, deploy_hash),
+    FOREIGN KEY (deploy_hash) REFERENCES deploys (hash)
```
One more key could be a `unique key (block_hash, deploy_position)`, which would automatically create an index as well.

Actually if the primary key was on `(block_hash, deploy_position)` and the table has `without rowid` then it's clustered in the order in which these deploys appear in the block and it never has to be reordered when we reconstruct a block. The unique key can be created on `(block_hash, deploy_hash)`, although strictly speaking it's up to the application to enforce this rule, so we may as well leave it out.

igor-ramazanov replied:
Done

aakoshh replied:
Hey, I'm sorry, I thought a bit more about it and I think we should have a unique index after all, and if we make it on `(deploy_hash, block_hash)` then quite possibly it can serve the queries where we do a `select * from deploy_process_results where deploy_hash=?`, given that the prefix column would be `deploy_hash`. 

The reason I think this would be a good idea is because we have to protect against replay attacks, so a block that has the same deploy twice would be wrong under all circumstances. Currently such a block would be invalid because of the nonce, they would change the same account so they couldn't commute and would be (hopefully) caught by validation, which happens before we insert into the database. But if we remove the nonce then we'll have to protect it in a different way, and this could be a good sanity check.

igor-ramazanov replied:
No problem, that was my initial thinking as well when I used `PRIMARY KEY` on `(deploy_hash, block_hash)`. Done

aakoshh commented:
> Yeah, that's a good idea. Though, I don't really know what's the best way to store block headers, because if I store them as a new column then it will be duplicated for all deploy process results which is highly inefficient from the storage space perspective.

Sorry, what I said was confusing. I was thinking, to avoid storing a big blob in SQLite we could store blocks like this:
```
table blocks(hash, validator, rank, timestamp, ..., header_data)
table deploy_process_results(block_hash, deploy_hash, position, ...)
table deploys(hash, account, timestamp, ...)
```
So I didn't mean to put the header into the deploy, I meant that if there's an orphaned block then all the deploys won't be stored again like they are now, the new block that re-proposes the old deploys just stores a new header and a new set of process results, referencing the existing deploy records.

aakoshh commented:
> Please, pay attention that I've decided to store deploy processing results for all blocks, even invalid and not sure if it's a good idea.

Since [this](https://github.com/CasperLabs/CasperLabs/commit/80b890b7c14ea6b7eccea65baec0cbeee04d7999) commit we are not storing invalid blocks. This is what Andreas suggested, these are blocks which are outright wrong. I can't remember what was the problem with storing them. This is not the same as equivocations, which are two valid but conflicting blocks from the same validator.

igor-ramazanov commented:
@aakoshh 
>Since this commit we are not storing invalid blocks. This is what Andreas suggested, these are blocks which are outright wrong. I can't remember what was the problem with storing them. This is not the same as equivocations, which are two valid but conflicting blocks from the same validator.

If we don't store invalid block, then it probably doesn't make a sense to store deploys process results from them because they won't be tied to any block. So, I'm going to change it aligning with blockstore behaviour.

aakoshh commented on storage/src/main/resources/db/migration/V20190808_666__Create_deploys_process_results_table.sql:
```diff
@@ -0,0 +1,37 @@
+CREATE TABLE deploys_process_results
```
I just noticed the name of this table; don't you think `deploy_` (singular) would be better?

igor-ramazanov replied:
I don't know, honestly, but if you are asking then probably it is :)

Done

igor-ramazanov commented:
@aakoshh Added a commit to not store process results from invalid blocks
